xr_sample_project/Assets/SampleViewer/Samples/GeoSpatialAR/Scripts/ArcGISGeospatialController.cs

```diff
@@ -0,0 +1,163 @@
+            }
+        }
+    }
+
+    private void SetLocation()
+    {
+        location = EarthManager.CameraGeospatialPose;
+        SetOrigin(new ArcGISPoint(location.Longitude, location.Latitude, 0, ArcGISSpatialReference.WGS84()));
+        Debug.Log("Origin Updated to Lat: " + location.Latitude + " Long: " + location.Longitude);
+    }
+
+    private void SetOriginLocation()
+    {
+        var earthTrackingState = EarthManager.EarthTrackingState;
+        
+        if (earthTrackingState == TrackingState.Tracking)
+        {
+            var cameraGeospatialPose = EarthManager.CameraGeospatialPose;
+            XROrigin.GetComponent<ArcGISLocationComponent>().Position = new ArcGISPoint(cameraGeospatialPose.Longitude, cameraGeospatialPose.Latitude, ArcGISSpatialReference.WGS84());
+            Camera.main.transform.position = Vector3.zero;
+            
+            CancelInvoke(nameof(SetOriginLocation));
+            Debug.LogWarning("Cancelled");
+        }
+    }
+
+    private void SetOrigin(ArcGISPoint OriginPoint)
+    {
+        mapComponent.OriginPosition = OriginPoint;
+    }
+
+    private void SetCamera(ArcGISPoint OriginPoint)
+    {
+        mapComponent.GetComponentInChildren<ArcGISCameraComponent>().gameObject.GetComponent<ArcGISLocationComponent>().Position = OriginPoint;
+    }
+    
+    private void AvailabilityCheck()
+    {
+        if (Input.location.status != LocationServiceStatus.Running)
+        {
+            Debug.LogWarning(
+                "Location services aren't running. VPS availability check is not available.");
+            return;
+        }
+
+        var location = Input.location.lastData;
+        var vpsAvailabilityPromise =
+            AREarthManager.CheckVpsAvailabilityAsync(location.latitude, location.longitude);
+        
+        if (vpsAvailabilityPromise.Result == VpsAvailability.Available)
+        {
+            precisionMarker.SetActive(false);
+        }
+        else
+        {
+            precisionMarker.SetActive(true);
+        }
+
+        Debug.LogFormat("VPS Availability at ({0}, {1}): {2}",
+            location.latitude, location.longitude, vpsAvailabilityPromise.Result);
+    }
+
+    private IEnumerator StartLocationService()
+    {
+        _waitingForLocationService = true;
+#if UNITY_ANDROID
+        if (!Permission.HasUserAuthorizedPermission(Permission.FineLocation))
+        {
+            Debug.Log("Requesting the fine location permission.");
+            Permission.RequestUserPermission(Permission.FineLocation);
+            yield return new WaitForSeconds(3.0f);
+        }
+#endif
+
+        if (!Input.location.isEnabledByUser)
+        {
+            Debug.Log("Location service is disabled by the user.");
+            _waitingForLocationService = false;
+            yield break;
+        }
+
+        Debug.Log("Starting location service.");
+        Input.location.Start();
+
+        while (Input.location.status == LocationServiceStatus.Initializing)
```

Comment:
Do we have any event we can hook into for this status change? While loop here is dangerous with no timeout.

Reply:
Unfortunately no. Location Services is still under old input and isn't available in new input. So there aren't any events. A lot of this code is just safety precautions.